File: src/main/java/frc/robot/subsystems/vision/VisionIO.java
package frc.robot.subsystems.vision;

import edu.wpi.first.math.geometry.Pose3d;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.geometry.Transform3d;
import org.littletonrobotics.junction.AutoLog;

/** AdvantageKit IO layer for vision measurements. */
public interface VisionIO {
  @AutoLog
  class VisionIOInputs {
    public boolean isConnected;
    public TargetObservation latestTargetObservation =
        new TargetObservation(new Rotation2d(), new Rotation2d());
    public PoseObservation[] poseObservations = new PoseObservation[0];
    public int[] tagIds = new int[0];
    public TargetTransform[] targetTransforms = new TargetTransform[0];
  }

  record TargetObservation(Rotation2d tx, Rotation2d ty) {}

  record PoseObservation(
      double timestamp,
      Pose3d pose,
      double ambiguity,
      int tagCount,
      double averageTagDistance) {}

  record TargetTransform(
      double timestamp,
      int fiducialId,
      Transform3d cameraToTarget,
      double ambiguity,
      double distance) {}

  default void updateInputs(VisionIOInputs inputs) {}
}

File: src/main/java/frc/robot/subsystems/vision/VisionIO.kt
package frc.robot.subsystems.vision

import edu.wpi.first.math.geometry.Pose3d
import edu.wpi.first.math.geometry.Rotation2d
import org.littletonrobotics.junction.LogTable
import org.littletonrobotics.junction.inputs.LoggableInputs

interface VisionIO {
    class VisionInputs : LoggableInputs {
        var isConnected: Boolean = false
        var latestTargetObservation: TargetObservation = TargetObservation(Rotation2d(), Rotation2d())
        var poseObservations: Array<PoseObservation> = emptyArray()
        var tagIDs = IntArray(0)

        override fun toLog(table: LogTable) {
            table.put("isConnected", isConnected)
            table.put("latestTargetObservation", latestTargetObservation)
            table.put("poseObservations", *poseObservations)
            table.put("tagIDs", tagIDs)
       }

        override fun fromLog(table: LogTable) {
            isConnected = table.get("isConnected", isConnected)
            latestTargetObservation = table.get("latestTargetObservation", latestTargetObservation)
            poseObservations = table.get("poseObservations", *poseObservations)
            tagIDs = table.get("tagIDs", tagIDs)
        }
    }

    @JvmRecord
    data class TargetObservation(val tx: Rotation2d, val ty: Rotation2d)

    @JvmRecord
    data class PoseObservation(
        val timestamp: Double,
        val pose: Pose3d,
        val ambiguity: Double,
        val tagCount: Int,
        val averageTagDistance: Double,
    )

    fun updateInputs(inputs: VisionInputs) {}
}